File: MercuryAPI-1.31.3.36/c/src/samples/multireadasync.c
/**
 * Sample program that reads tags on multiple readers and prints the tags found.
 * @file multireadasync.c
 */

#include <tm_reader.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <inttypes.h>
#include <string.h>
#ifndef WIN32
#include <unistd.h>
#endif

/* Enable this to use transportListener */
#ifndef USE_TRANSPORT_LISTENER
#define USE_TRANSPORT_LISTENER 0
#endif

#define usage() {errx(1, "Please provide valid reader URL, such as: reader1-uri [--ant n] reader2-uri [--ant n]\n"\
                         "reader-uri : e.g., 'tmr:///COM1' or 'tmr:///dev/ttyS0/' or 'tmr://readerIP'\n"\
                         "[--ant n] : e.g., '--ant 1'\n"\
                         "Example: 'tmr:///com4 tmr:///com13' or 'tmr:///com4 --ant 1 tmr:///com13 --ant 2' \n");}

typedef struct antennaValues
{
  uint8_t antennaBuffer[20];
  uint8_t antennaCount;
} antennaValues;

typedef struct readerDesc
{
  char uri[TMR_MAX_READER_NAME_LENGTH];
  int idx;
  TMR_ReadPlan *readPlan;
  antennaValues _antStorage;
} readerDesc;

void errx(int exitval, const char *fmt, ...)
{
  va_list ap;

  va_start(ap, fmt);
  vfprintf(stderr, fmt, ap);

  exit(exitval);
}

void checkerr(TMR_Reader* rp, TMR_Status ret, int exitval, const char *msg)
{
  if (TMR_SUCCESS != ret)
  {
    errx(exitval, "Error %s: %s\n", msg, TMR_strerr(rp, ret));
  }
}

void serialPrinter(bool tx, uint32_t dataLen, const uint8_t data[],
                   uint32_t timeout, void *cookie)
{
  FILE *out = stdout;
  readerDesc *rdp = cookie;
  uint32_t i;

  fprintf(out, "%s %s", rdp->uri, tx ? "Sending: " : "Received:");
  for (i = 0; i < dataLen; i++)
  {
    if (i > 0 && (i & 15) == 0)
      fprintf(out, "\n         ");
    fprintf(out, " %02x", data[i]);
  }
  fprintf(out, "\n");
}

void stringPrinter(bool tx,uint32_t dataLen, const uint8_t data[],uint32_t timeout, void *cookie)
{
  FILE *out = stdout;

  fprintf(out, "%s", tx ? "Sending: " : "Received:");
  fprintf(out, "%s\n", data);
}

void parseAntennaList(uint8_t *antenna, uint8_t *antennaCount, char *args)
{
  char *token = NULL;
  char *delimiters = ",";
  uint8_t i = 0;
  int scans;

  /* get the first token */
  if (NULL == args)
  {
    fprintf(stdout, "Missing argument\n");
    usage();
  }

  token = strtok(args, delimiters);
  if (NULL == token)
  {
    fprintf(stdout, "Can't parse '%s' as comma-separated list\n", args);
    usage();
  }

  while(NULL != token)
  {
    scans = sscanf(token, "%"SCNu8, &antenna[i]);
    if (1 != scans)
    {
      fprintf(stdout, "Can't parse '%s' as an 8-bit unsigned integer value\n", token);
      usage();
    }
    i++;
    token = strtok(NULL, delimiters);
  }
  *antennaCount = i;
}

void callback(TMR_Reader *reader, const TMR_TagReadData *t, void *cookie);
void exceptionCallback(TMR_Reader *reader, TMR_Status error, void *cookie);

int main(int argc, char *argv[])
{

#ifndef TMR_ENABLE_BACKGROUND_READS
  errx(1, "This sample requires background read functionality.\n"
          "Please enable TMR_ENABLE_BACKGROUND_READS in tm_config.h\n"
          "to run this codelet\n");
  return -1;
#else

  TMR_Reader *r;
  readerDesc *rd;
  int rcount;
  TMR_Reader *rp;
  TMR_Status ret;
  TMR_Region region;
  TMR_ReadListenerBlock *rlb;
  TMR_ReadExceptionListenerBlock *reb;
  char *readerName = NULL;
  uint8_t *antennaList = NULL;
#if USE_TRANSPORT_LISTENER
  TMR_TransportListenerBlock *tb;
#endif
  int i;

  if (argc < 2)
  {
    usage();
  }
  
  /* this is a non-optimized, worst-case estimate. */
  rcount = argc-1;
  r = (TMR_Reader*) calloc(rcount, sizeof(TMR_Reader));
  rd = (readerDesc*) calloc(rcount, sizeof(readerDesc));
  rlb = (TMR_ReadListenerBlock*) calloc(rcount, sizeof(TMR_ReadListenerBlock));
  reb = (TMR_ReadExceptionListenerBlock*) calloc(rcount, sizeof(TMR_ReadExceptionListenerBlock));
  
  rd->idx = 0;
  for (i = 1; i < argc; i++)
  {
    if(0 == strcmp("--ant", argv[i]))
    {
      /* Its a antenna list */
      if (NULL != readerName)
      {
        if (NULL != antennaList)
        {
          fprintf(stdout, "Duplicate argument: --ant specified more than once\n");
          usage();
        }
        parseAntennaList(rd[rd->idx -1]._antStorage.antennaBuffer, &rd[rd->idx -1]._antStorage.antennaCount, argv[i+1]);
        i++;
        antennaList = rd[rd->idx -1]._antStorage.antennaBuffer;
      }
      else
      {
        usage();
      }
    }
    else
    {
      /* Its a reader name */      
      strcpy(rd[rd->idx].uri, argv[i]);
      readerName = rd[rd->idx].uri;
      rd->idx++;
      antennaList = NULL;
    }
  }

  for (i = 0; i < rd->idx; i++)
  {
    rp = &r[i];
    ret = TMR_create(rp, rd[i].uri);
    checkerr(rp, ret, 1, "creating reader %s");

    printf("Created reader %d: %s\n", i+1, rd[i].uri);

#if USE_TRANSPORT_LISTENER

    tb = (TMR_TransportListenerBlock*) calloc(rcount, sizeof(TMR_TransportListenerBlock));

    if (TMR_READER_TYPE_SERIAL == rp->readerType)
    {
      tb[i].listener = serialPrinter;
    }
    else
    {
      tb[i].listener = stringPrinter;
    }
    tb[i].cookie = &rd[i];

    TMR_addTransportListener(rp, &tb[i]);
#endif


    //TMR_SR_PowerMode pm = TMR_SR_POWER_MODE_FULL;
    //ret = TMR_paramSet(rp, TMR_PARAM_POWERMODE, &pm);

    ret = TMR_connect(rp);
    checkerr(rp, ret, 1, "connecting reader");

    region = TMR_REGION_NONE;
    ret = TMR_paramGet(rp, TMR_PARAM_REGION_ID, &region);
    checkerr(rp, ret, 1, "getting region");

    if (TMR_REGION_NONE == region)
    {
      TMR_RegionList regions;
      TMR_Region _regionStore[32];
      regions.list = _regionStore;
      regions.max = sizeof(_regionStore)/sizeof(_regionStore[0]);
      regions.len = 0;

      ret = TMR_paramGet(rp, TMR_PARAM_REGION_SUPPORTEDREGIONS, &regions);
      checkerr(rp, ret, __LINE__, "getting supported regions");

      if (regions.len < 1)
      {
        checkerr(rp, TMR_ERROR_INVALID_REGION, __LINE__, "Reader doesn't support any regions");
      }
      region = regions.list[0];
      ret = TMR_paramSet(rp, TMR_PARAM_REGION_ID, &region);
      checkerr(rp, ret, 1, "setting region");  
    }
  /**
   * Checking the software version of the sargas.
   * The antenna detection is supported on sargas from software version of 5.3.x.x.
   * If the Sargas software version is 5.1.x.x then antenna detection is not supported.
   * User has to pass the antenna as arguments.
   */
  {
    ret = isAntDetectEnabled(rp, &(rd[i]._antStorage.antennaCount));
    if(TMR_ERROR_UNSUPPORTED == ret)
    {
      fprintf(stdout, "Reader doesn't support antenna detection. Please provide antenna list.\n");
      usage();
    }
    else
    {
      checkerr(rp, ret, 1, "Getting Antenna Detection Flag Status");
    }
  }
    /**
    * for antenna configuration we need two parameters
    * 1. antennaCount : specifies the no of antennas should
    *    be included in the read plan, out of the provided antenna list.
    * 2. antennaList  : specifies  a list of antennas for the read plan.
    **/ 

    // initialize the read plan 
    rd[i].readPlan = (TMR_ReadPlan*) malloc(sizeof(TMR_ReadPlan));
    if (rd[i]._antStorage.antennaCount)
    {
      ret = TMR_RP_init_simple(rd[i].readPlan, rd[i]._antStorage.antennaCount, rd[i]._antStorage.antennaBuffer, TMR_TAG_PROTOCOL_GEN2, 1000);
      checkerr(rp, ret, 1, "initializing the  read plan");
    }
    else
    {
      ret = TMR_RP_init_simple(rd[i].readPlan, rd[i]._antStorage.antennaCount, NULL, TMR_TAG_PROTOCOL_GEN2, 1000);
      checkerr(rp, ret, 1, "initializing the  read plan");
    }
    

    /* Commit read plan */
    ret = TMR_paramSet(rp, TMR_PARAM_READ_PLAN, rd[i].readPlan);
    checkerr(rp, ret, 1, "setting read plan");

    rlb[i].listener = callback;
    rlb[i].cookie = &rd[i];

    reb[i].listener = exceptionCallback;
    reb[i].cookie = NULL;

    ret = TMR_addReadListener(rp, &rlb[i]);
    checkerr(rp, ret, 1, "adding read listener");

    ret = TMR_addReadExceptionListener(rp, &reb[i]);
    checkerr(rp, ret, 1, "adding exception listener");

    ret = TMR_startReading(rp);
    checkerr(rp, ret, 1, "starting reading");
  }

#ifndef WIN32
  sleep(5);
#else
  Sleep(5000);
#endif

  for (i=0; i<rd->idx; i++)
  {
    rp = &r[i];
    ret = TMR_stopReading(rp);
    checkerr(rp, ret, 1, "stopping reading");
    TMR_destroy(rp);
  }
  return 0;

#endif /* TMR_ENABLE_BACKGROUND_READS */
}


void
callback(TMR_Reader *reader, const TMR_TagReadData *t, void *cookie)
{
  char epcStr[128];
  readerDesc *rdp = cookie;

  TMR_bytesToHex(t->tag.epc, t->tag.epcByteCount, epcStr);
  printf("%s: %s\n", rdp->uri, epcStr);
}

void
exceptionCallback(TMR_Reader *reader, TMR_Status error, void *cookie)
{
  fprintf(stdout, "Error:%s\n", TMR_strerr(reader, error));
}

File: MercuryAPI-1.31.3.36/java/samples_nb/src/samples/multireadasync.java
/**
 * Sample program that reads tags in the background and prints the tags found.
 */
// Import the API
package samples;

import com.thingmagic.*;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Level;

public class multireadasync
{

    static void usage()
    {
        System.out.printf("Usage: Please provide valid arguments, such as:\n"
                + "multireadasync [-v] [reader-uri1] [--ant n[,n...]] [reader-uri2] [--ant n[,n...]]  \n" +
                  "-v  Verbose: Turn on transport listener\n" +
                  "reader-uri  Reader URI: e.g., \"tmr:///COM1\", \"tmr://astra-2100d3\"\n"
                + "--ant  Antenna List: e.g., \"--ant 1\", \"--ant 1,2\"\n" +
                  " e.g:'tmr:///COM1 --ant 1,2 tmr:///COM2 --ant 1,2 or 'tmr:///COM1 --ant 1,2 tmr://10.11.115.36 --ant 1,2\n");
        System.exit(1);
    }

    public static void main(String argv[])
    {

        // Create Reader object, connecting to physical device
        try
        {
            int[] antennaList = null;
            String readerName = null;
            Map<String, int[]> readerPort = new HashMap<String, int[]>();           
            TagReadData[] tags;

            for (int nextarg = 0; nextarg < argv.length; nextarg++)
            {
                String arg = argv[nextarg];
                if (arg.equalsIgnoreCase("--ant"))
                {
                    if (antennaList != null)
                    {
                        System.out.println("Duplicate argument: --ant specified more than once");
                        usage();
                    }
                    antennaList = parseAntennaList(argv, nextarg);
                    nextarg++;
                }
                else
                {
                    if (readerName != null)
                    {
                        readerPort.put(readerName, antennaList);
                        antennaList = null;
                    }
                    readerName = arg;
                }
            }

            if (readerName != null)
            {
                readerPort.put(readerName, antennaList);
                readerName = null;
                antennaList = null;
            }
           
            Reader[] r = new Reader[readerPort.size()];
            int i =0;
            for (String reader: readerPort.keySet())
            {
                r[i] = Reader.create(reader);
                r[i].connect();
                if (Reader.Region.UNSPEC == (Reader.Region) r[i].paramGet("/reader/region/id"))
                {
                    Reader.Region[] supportedRegions = (Reader.Region[]) r[i].paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
                    if (supportedRegions.length < 1)
                    {
                        throw new Exception("Reader doesn't support any regions");
                    }
                    else
                    {
                        r[i].paramSet("/reader/region/id", supportedRegions[0]);
                    }
                }

                if (r[i].isAntDetectEnabled(antennaList) && readerPort.get(reader) == null)
                {
                    System.out.println("Module doesn't has antenna detection support, please provide antenna list");
                    r[i].destroy();
                    usage();
                }
                
                SimpleReadPlan plan = new SimpleReadPlan(readerPort.get(reader), TagProtocol.GEN2, null, null, 1000);
                r[i].paramSet(TMConstants.TMR_PARAM_READ_PLAN, plan);
               
                // Create and add tag listener
                ReadListener rl = new PrintListener();
                r[i].addReadListener(rl);

                ReadExceptionListener exceptionListener = new TagReadExceptionReceiver();
                r[i].addReadExceptionListener(exceptionListener);

                // search for tags in the background
                r[i].startReading();
                i++;
            }
            Thread.sleep(1000);

            for (i = 0; i < readerPort.size(); i++)
            {
                r[i].stopReading();
                // Shut down reader
                r[i].destroy();
            }

        }
        catch (ReaderException re)
        {
            System.out.println("ReaderException: " + re.getMessage());
        }
        catch (Exception re)
        {
            System.out.println("Exception: " + re.getMessage());
        }
    }

    static class PrintListener implements ReadListener
    {

        public void tagRead(Reader r, TagReadData tr)
        {
            try
            {
                System.out.println(r.paramGet("/reader/uri").toString()+" Background read: " + tr.toString());
            }
            catch (ReaderException ex)
            {
                java.util.logging.Logger.getLogger(multireadasync.class.getName()).log(Level.SEVERE, null, ex);
            }
        }

    }

    static class TagReadExceptionReceiver implements ReadExceptionListener
    {

        public void tagReadException(com.thingmagic.Reader r, ReaderException re)
        {
            System.out.println("Reader Exception: " + re.getMessage());
            if (re.getMessage().equals("Connection Lost"))
            {
                System.exit(1);
            }
        }
    }

    static int[] parseAntennaList(String[] args, int argPosition)
    {
        int[] antennaList = null;
        try
        {
            String argument = args[argPosition + 1];
            String[] antennas = argument.split(",");
            int i = 0;
            antennaList = new int[antennas.length];
            for (String ant : antennas)
            {
                antennaList[i] = Integer.parseInt(ant);
                i++;
            }
        }
        catch (IndexOutOfBoundsException ex)
        {
            System.out.println("Missing argument after " + args[argPosition]);
            usage();
        }
        catch (Exception ex)
        {
            System.out.println("Invalid argument at position " + (argPosition + 1) + ". " + ex.getMessage());
            usage();
        }
        return antennaList;
    }
}
